docs/taskinputvalidation.md

```diff
@@ -0,0 +1,135 @@
+Task input validation
+=====================
+
+Goals
+-------
+
+1. Task authors should be able to do wellknown and custom validations with ease.
+    - Eg: A valid - URL, IP Address, e-email address, Date, Port number, Cert thumbprint, user name (IDN or UPN format), input based (length of input, valid `path` characters, case checkings)
+    - Let task authors provide any custom regex validations
+    - Note: To prevent [reDoS](https://www.owasp.org/index.php/Regular_expression_Denial_of_Service_-_ReDoS) we should give a [timeout](https://msdn.microsoft.com/en-us/library/hh160196%28v=vs.110%29.aspx) to regex validation.
+2. *Non goal/Goal to get holistic view* - Provide task input enforcement and also ability to advise possible input values
+    - Eg: If a task input called `destination` is `abc` then the possible list of values for a task input called `target` should be
+    *one of* `[x,y,z]`. This imples that the value of the input `target` should be enforced to be one of them *and* also the UI should advise those inputs as possible values to the user.
+3. *Non goal/Unclear goal* - Authors should be able to run custom javascript
+    - Task authors can point to a javascript file and render the whole task editor in iframe?
+    - Task authors can point to custom javascript to control just task input's visibility and validation?
+
+-------
+
+Supported boolean functions
+-------
+
+> `value` is alias to task input value
+
+- `isUrl(value: string)` - true if value is of valid URL, allows ftp urls as well
+- `isIpV4Address(value: string)` - true if value is a valid IPV4 address
+- `isEmail(value: string)` - true if value is of email format
+- `isEmpty(value: string)` - true if value is empty
+- `isInRange(value: string, min: number, max: number)` - true if value is <= max and >= min
+- `isSha1(value: string)` - true if value is a valid sha1 hash
+- `isWinUsername(value: string)` - true if value is of samAccountName or userPrincipalName format
+- `isLowerCase(value: string)` - true if value is all lower case
+- `isUpperCase(value: string)` - true if value is all upper case
+- `isMinlength(value: string, min: number)` - true if length is greater than or equal to number
```

Comment:
One strategy is how it reads. For example, consider `contains`. The parameters are arranged so that it logically reads "x contains y".

Another strategy is to accommodate N parameters. It looks like the former strategy is the predominant one and probably makes sense for most usage as well (e.g. atypical to apply same regex to many different values).